pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go

```diff
@@ -0,0 +1,313 @@
+			case "Lock":
+				lt.addLock(call, false)
+			case "RLock":
+				lt.addLock(call, true)
+			case "Unlock":
+				lt.maybeReportUnlock(call, false)
+			case "RUnlock":
+				lt.maybeReportUnlock(call, true)
+			}
+			return true
+		}
+	}
+
+	return false
+}
+
+// addLock adds a new LockInfo to the end of the locks slice.
+func (lt *LockTracker) addLock(call *ast.CallExpr, isRead bool) {
+	sel, ok := call.Fun.(*ast.SelectorExpr)
+	if !ok {
+		return
+	}
+	// We add the other fields when we traverse and reach the *ast.Ident
+	lt.locks = append(lt.locks, &LockInfo{
+		isRead:           isRead,
+		foundUnlockMatch: false,
+	})
+	ast.Inspect(sel, func(n ast.Node) bool {
+		ident, ok := n.(*ast.Ident)
+		if !ok {
+			return true
+		}
+		// If we find Lock or RLock set the line and we are done
+		// traversing.
+		if ident.Name == "Lock" || ident.Name == "RLock" {
+			position := lt.pass.Fset.Position(n.Pos())
+			lt.locks[len(lt.locks)-1].lineNum = position.Line
+			return false
+		}
+		// Otherwise its part of the prefix that we'll add.
+		lt.locks[len(lt.locks)-1].prefix += ident.Name + "."
+		// Keep traversing if we didn't find an *ast.Ident node.
+		return true
+	})
+}
+
+// maybeReportUnlock tries to find a matching lock for a given unlock by
+// iterating backwards in the locks slice. If one is found, it checks if the
+// distance between is greater than maxLineDistance and also has no nlint
+// comment and reports on that if both are true.
+func (lt *LockTracker) maybeReportUnlock(call *ast.CallExpr, isRead bool) {
+	sel, ok := call.Fun.(*ast.SelectorExpr)
+	if !ok {
+		return
+	}
+	var node ast.Node
+	var unlockPrefix string
+	var unlockLineNum int
+	var unlockPos token.Pos
+	ast.Inspect(sel, func(n ast.Node) bool {
+		ident, ok := n.(*ast.Ident)
+		if !ok {
+			return true
+		}
+		// If we find Unlock or RUnlock set the line and we are done
+		// traversing.
+		if ident.Name == "Unlock" || ident.Name == "RUnlock" {
+			position := lt.pass.Fset.Position(n.Pos())
+			unlockPos = n.Pos()
+			unlockLineNum = position.Line
+			node = n
+			return false
+		}
+		// Otherwise its part of the prefix that we'll add.
+		unlockPrefix += ident.Name + "."
+		// Keep traversing if we didn't find an *ast.Ident node.
+		return true
+	})
+	// Reverse iterate through the locks slice searching for a match.
+	for i := len(lt.locks) - 1; i >= 0; i-- {
+		lock := lt.locks[i]
+		if lock.foundUnlockMatch {
+			continue
+		}
+		// It's only an issue if prefixes match, the lock type matches, distance between is >5 lines away and
+		// there is no nolint comment.
+		if unlockPrefix == lock.prefix && isRead == lock.isRead {
+			lockDistance := unlockLineNum - lock.lineNum
+			if lockDistance > maxLineDistance && !passesutil.HasNolintComment(lt.pass, node, noLintName) {
+				lt.issues = append(lt.issues, analysis.Diagnostic{
+					Pos:     unlockPos,
+					Message: fmt.Sprintf("{RW}Mutex {R}Unlock is >%d lines away from matching {R}Lock", maxLineDistance),
```

Comment:
The error message here (and on line 311) is not great. Note that this is the message that will be displayed to devs when they violate the lint rule. In particular it's unclear what someone is supposed to do to solve the problem. e.g. "Oh, I can't have more than 5 lines between the lock and unlock, so I need to start combining multiple lines of code into one line with semicolons" or something like that.

The error message should succinctly describe what the actual problem is and suggest a fix like "move the `Unlock()` call to a `defer` statement immediately after the `Lock()` call".

The error message on line 311 has an additional problem because what is "non-linear control flow"? Why is it a problem if I have non-linear control flow, and how do I fix it?